azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+
+        # If we get a request to disconnect, we either complete the request immediately
+        # (if we're already disconencted) or we pass the disconnect request down.
+        elif op.name is pipeline_ops.names.Iot_Disconnect:
+            if not self.connected:
+                logger.info("transport is already disconencted.  completing early.")
+                self.complete_op(op)
+            else:
+                self.continue_op(op)
+
+        # Any other operation that requires a connection can trigger a connection if i
+        # we're not connected.
+        elif op.needs_connection and not self.connected:
+            self._do_connect(op)
+
+        # Finally, if this stage doesn't need to do anything else with this operation,
+        # it just passes it down.
+        else:
+            self.continue_op(op)
+
+    def _block(self, op):
+        """
+        block this stage while we're waiting for the connection to complete.
+        """
+        logger.info("{}({}) blocking".format(self.name, op.name))
+        self.blocked = True
+
+    def _unblock(self, op, error):
+        """
+        Unblock this stage after the connection is complete.  This also means
+        releasing all the queued up operations that we were waiting for the
+        connect operation to complete.
+        """
+        logger.info("{}({}) unblocking".format(self.name, op.name))
+        self.blocked = False
+        logger.info(
+            "{}({}) processing {} items in queue".format(self.name, op.name, len(self.queue))
+        )
+        # loop through our queue and release all the blocked operations
+        while len(self.queue):
+            op_to_release = self.queue.popleft()
+            # if we're unblocking the queue because something (like a connect operation) failed,
+            # then we fail all of the blocked operations with the same error.
+            if error:
+                op_to_release.error = error
+                self.complete_op(op_to_release)
+            else:
+                # when we release, go back through this stage again to make sure requirements are _really_ satisfied.
+                # this also pre-maturely completes ops that might now be satisfied.
+                self.run_op(op_to_release)
+
+    def _do_connect(self, op=None):
+        """
+        Start connecting the transport in response to some operation (which may or may not be an Iot_Connect operation)
+        """
+        # first, we block the queue while we're connecting
+        self._block(op=op)
+
+        # If we're connecting as a side-effect of some other operation (that is not Iot_Connect), then we queueu
+        # that operation to run after the connection is complete.
+        if op.name is not pipeline_ops.names.Iot_Connect:
+            logger.info("{}({}) queueing until connection complete".format(self.name, op.name))
+            self.queue.append(op)
+
+        # callback that gets called after we're connected.
```

Comment:
>callback [](start = 10, length = 8)

as always, you know how i feel about things referred to as callbacks before they are provided to a consumer